test/test_producer.py

```diff
@@ -40,3 +51,147 @@ def partitions(topic):
         topic = b"test-topic"
         producer.send_messages(topic, b'hi')
         assert client.send_produce_request.called
+
+    def test_producer_async_queue_overfilled_batch_send(self):
+        queue_size = 2
+        producer = Producer(MagicMock(), batch_send=True,
+                            async_queue_maxsize=queue_size)
+
+        topic = b'test-topic'
+        partition = 0
+        message = b'test-message'
+
+        with self.assertRaises(AsyncProducerQueueFull):
+            message_list = [message] * (queue_size + 1)
+            producer.send_messages(topic, partition, *message_list)
```

Comment:
@vshlapakov you can see how we did that in our fork at https://github.com/docker-hub/kafka-python/blob/master/test/test_producer.py
It seemed like verifying that 2 messages went through called for a race condition, so we mocked-out the `_send_upstream` call and drained the queue so the thread wouldn't hang around. 
That may help in case you start seeing spurious errors.